ouroboros-consensus-cardano/src/unstable-cardano-tools/Cardano/Tools/Headers.hs

```diff
@@ -0,0 +1,89 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- | Tooling to generate and validate (Praos) headers.
+module Cardano.Tools.Headers where
+
+import Cardano.Crypto.DSIGN (deriveVerKeyDSIGN)
+import Cardano.Crypto.VRF (VRFAlgorithm (deriveVerKeyVRF, hashVerKeyVRF))
+import Cardano.Ledger.Api (ConwayEra, StandardCrypto)
+import Cardano.Ledger.BaseTypes (BoundedRational (boundRational), PositiveUnitInterval, mkActiveSlotCoeff)
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Compactible (toCompact)
+import Cardano.Ledger.Keys (VKey (..), hashKey)
+import Cardano.Ledger.PoolDistr (IndividualPoolStake (..))
+import Cardano.Protocol.TPraos.OCert (ocertN)
+import Control.Monad.Except (runExcept)
+import qualified Data.Aeson as Json
+import qualified Data.ByteString.Lazy as LBS
+import Data.Char (isSpace)
+import qualified Data.Map as Map
+import Data.Maybe (fromJust)
+import qualified Data.Text as Text
+import Ouroboros.Consensus.Block (validateView)
+import Ouroboros.Consensus.Protocol.Praos (
+    Praos,
+    doValidateKESSignature,
+    doValidateVRFSignature,
+ )
+import Ouroboros.Consensus.Protocol.Praos.Header (Header, hbOCert, pattern Header)
+import Ouroboros.Consensus.Shelley.HFEras ()
+import Ouroboros.Consensus.Shelley.Ledger (
+    ShelleyBlock,
+    mkShelleyHeader,
+ )
+
+import Ouroboros.Consensus.Shelley.Protocol.Praos ()
+import Test.Ouroboros.Consensus.Protocol.Praos.Header (
+    GeneratorContext (..),
+    MutatedHeader (..),
+    Mutation (..),
+    Sample (..),
+    expectedError,
+    generateSamples,
+    header,
+    mutation,
+ )
+
+type ConwayBlock = ShelleyBlock (Praos StandardCrypto) (ConwayEra StandardCrypto)
+
+-- * Running Generator
+data Options = Options
+
+run :: Options -> IO ()
+run Options = do
+    sample <- generateSamples
+    LBS.putStr $ Json.encode sample <> "\n"
+
+data ValidationResult = Valid Mutation | Invalid Mutation String
+    deriving (Eq, Show)
+
+validate :: GeneratorContext -> MutatedHeader -> ValidationResult
```

Comment:
Do you have a sense yet of how difficult it would be to reuse the implementation's validation logic instead of re-specifying it here? Maybe we could do some rearranging upstream (eg factor out a "worker") to make it easier to invoke in this incomplete calling context.

Unless you were intending for this code to be independent?

Reply:
Esgen also mentioned that maybe the Agda spec would be amenable to reuse here?

Reply:
EG you're missing the `chainChecks` and the opcert checks. But I suppose that makes sense considering you're not so much interested in the actual incoming ledger/header state.

But maybe your generator could also generate a header state and the ledger view alongside each header? IE all three ingredients necessary to invoke `Ouroboros.Consensus.HeaderValidation.validateHeader`.

Reply:
I agree the `validate` function in its current form is cumbersome, but ultimately it should just be a thin wrapper over calling the VRF and KES validation functions from `Praos` module. Creating the necessary environment to directly call `validateHeader` seemed like a more ambitious task at this stage, and something that could be done in a later PR to avoid "scope creep" as there really are 2 independent stages in the header validation.

re Agda spec: that's definitely something I think would be interesting to reuse as testing against an independent specification of the validation logic makes a lot of sense. This means the specification should be modular enough and separate the validation steps perhaps?

Reply:
> re Agda spec: that's definitely something I think would be interesting to reuse as testing against an independent specification of the validation logic makes a lot of sense. This means the specification should be modular enough and separate the validation steps perhaps?

The spec is structured in STS style, with a DAG of rules, see e.g. Figure 6 in the Agda spec [here](https://ouroboros-consensus.cardano.intersectmbo.org/docs/about-ouroboros/References). I guess that when exposing the spec as runnable Haskell functions (which I think shouldn't be *that* far off after #1270), this could be modular to e.g. have one function per STS sub-rule.[^1]

While being more granular that what is present in the current implementation, there is not directly something that corresponds exactly to VRF + KES validation. OCERT corresponds to KES validation, and PRTCL does VRF validation, but also a few more things like nonce evolution (not very complicated).

[^1]: (Of course, we could also structure the existing Haskell code in such a way that different STS sub-rules can be invoked independently.

Reply:
Thanks for the links. Of course, modelling the tests after the specification's transitions or logic rules would be desirable. BTW, having a hyperlinked HTML rendering of the Agda spec would be much more convenient than a PDF. Or did I miss the link somehow?

Reply:
> BTW, having a hyperlinked HTML rendering of the Agda spec would be much more convenient than a PDF. Or did I miss the link somehow?

Good point, we don't have that ATM, might to it later when I find the time :+1: